### ISSUE TITLE
Error when detaching integration methods

### ISSUE THREAD

**Issue by joaander** [glotzerlab/hoomd-blue#998]
## Description

MD integration methods cannot be removed from the methods list after attaching.

## Script

```python
# Include a minimal script that reproduces the problem
import hoomd

snap = hoomd.Snapshot()
snap.particles.N = 2
snap.particles.types = ['A']
sim = hoomd.Simulation(hoomd.device.CPU(), seed=1)
sim.create_state_from_snapshot(snap)

integrator = hoomd.md.Integrator(dt=0.005)
nvt = hoomd.md.methods.Langevin(kT=1.5, filter=hoomd.filter.All())
integrator.methods.append(nvt)
sim.operations.integrator = integrator

sim.run(0)
integrator.methods.pop(0)
```

## Output

```
Traceback (most recent call last):
  File "/scratch/joaander/test/detach-method.py", line 15, in <module>
    integrator.methods.pop(0)
  File "/usr/lib/python3.9/_collections_abc.py", line 1100, in pop
    del self[index]
  File "/scratch/joaander/build/hoomd/hoomd/data/syncedlist.py", line 153, in __delitem__
    self._list[index]._detach()
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 233, in _detach
    self._update_param_dict()
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 279, in _update_param_dict
    self._param_dict[key] = getattr(self, key)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 46, in __getattr__
    return self._getattr_param(attr)
  File "/scratch/joaander/build/hoomd/hoomd/md/methods.py", line 82, in _getattr_param
    parameter = getattr(self._cpp_obj, attr)
AttributeError: 'hoomd.md._md.TwoStepLangevin' object has no attribute 'tolerance'
```

## Expected output

No output.

## Configuration

Platform:
- Linux
- CPU

Installation method:
- Compiled from source

Versions:
- Python version: 3.9
- **HOOMD-blue** version: master branch

## Developer

@b-butler do you have any ideas?

**Comment by b-butler** [comment]
@joaander, yes I believe I know the case. It is somewhat subtle, but since the rattle methods have tolerances and manifolds, but the non-rattle one's do not. In `_update_param_dict` the Python side is querying for those quantities (since they are defined on the Python side), but they don't exist on the C++ side. A few possible solutions I briefly thought of

1. Catch these `AttribtueError`s in `_update_param_dict` and assume they all come from cases like this.
2. Add a hook for potentially skipping parameters in `_update_param_dict`.
3. Add a `_update_param_dict` or don't call `super()` in `hoomd.md.methods.MethodRATTLE`.
4. In the C++ pybind11 class definition create dummy tolerances and manifold_constraints e.g.
``` C++
.def("tolerance", [](){ return 1e-5; }, [](const auto& tolerance) { })
```

**Comment by vyasr** [comment]
Would it make more sense to just separate the RATTLE and non-RATTLE classes (maybe via inheritance) on the Python side? The param dict infrastructure is built around the idea of synchronization between C++ and Python codes that represent the same object. It seems conceptually undesirable to introduce mechanisms that allow them to break the expected mirroring unless there's a very good reason to do so.

**Comment by b-butler** [comment]
@vyasr that is a good perspective. I originally advocated for the combination of them into one Python class as it seemed to me that from the user perspective the RATTLE methods are just the same integrator with another feature, so that could be hidden into a single Python class.

In general, I tried to design much of the API to have hooks that allow for varied behavior. However, I don't know if I intended `_update_param_dict` to serve that function (I likely did not). So from an internal API side the suggests do somewhat break the mold.

Ultimately, it comes down to how much does combining the C++ classes into a single Python class help the user versus how much does maintaining paradigms on the developer side aid development and code maintenance.

As I side note, I need to document the `Operation`/`_HOOMDBaseObject` API more thoroughly for other developers :sweat_smile:
